aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -515,6 +515,369 @@ Tensor& matmul_out(Tensor &result, const Tensor & tensor1, const Tensor & tensor
   return result;
 }
 
+// helper methods for matrix_exp
+namespace {
+
+template <typename scalar_t, int ROW, int COL>
+using array2d = std::array<std::array<scalar_t, COL>, ROW>;
+
+// we consider 6 Taylor expansions of degree
+// 1, 2, 4, 8, 12, 18
+constexpr int total_n_degs = 6;
+
+// 0! to 4!
+// Introduced to avoid 'magic number's clang-tidy complaints about
+constexpr int fact_array_size = 5;
+constexpr std::array<float, fact_array_size> fact = {1., 1., 2., 6., 24.};
+
+// Iterates over first `n_batch_dims` and applies `f` to each element.
+// TODO: this implemtation is not efficient, it simply iteratates over
+// batch elements in a serial for-loop.
+template <typename func_t>
+void batch_apply(at::ArrayRef<Tensor> tensors, int64_t n_batch_dims, const func_t& f) {
+  // no-op if batch is empty
+  if (n_batch_dims <= 0) {
+    return;
+  }
+
+  std::vector<Tensor> tensors_view;
+  for (auto& tensor : tensors) {
+    std::vector<int64_t> tensor_view_sizes = {-1};
+    for (auto d = n_batch_dims; d < tensor.dim(); ++d) {
+      tensor_view_sizes.push_back(tensor.size(d));
+    }
+    auto tensor_view = tensor.view(tensor_view_sizes);
+    tensors_view.push_back(tensor_view);
+  }
+
+  for (int64_t input_idx = 0; input_idx < tensors_view[0].size(0);
+    ++input_idx) {
+    std::vector<Tensor> input_selected;
+    for (auto& tensor_view : tensors_view) {
+      auto curr_input = tensor_view.select(0, input_idx);
+      input_selected.push_back(curr_input);
+    }
+    f(input_selected);
+  }
+}
+
+Tensor matrix_power(const Tensor& matrices, const Tensor& powers) {
```

Comment:
Since this is a new variant of `matrix_power`, this should be documented. It seems like this is not bound to the Python interface.

Also, is `powers` a 1-dim `Tensor`? There should be a check for this, if so.

Reply:
This `matrix_power` is not publicly exposed, used only for `matrix_exp` so far.

Reply:
The method is not publicly available, hence `powers` if of right shape. Even if `powers` is 1-dim tensor, it will work, as reshaping takes places by substituting the first dimension by `-1`, which does not change anything in this case...